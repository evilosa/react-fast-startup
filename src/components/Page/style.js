export const PageColors = {
  color: '#fff',
  background: '#222222',
};

export const createPageStyle = (/*colors = PageColors*/) => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
    width: '100%',
    height: '100%',
  },

  header: {
    display: 'flex',
    flex: '1 0',
    height: 'auto',
  },

  content: {
    display: 'flex',
    flexDirection: 'column',
    flex: '1 0',
  },

  footer: {
    display: 'flex',
    flex: '1 0',
    height: 'auto',
  }
});

export const ButtonStyle = createPageStyle();

export default ButtonStyle;
