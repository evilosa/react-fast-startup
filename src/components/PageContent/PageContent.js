import React from 'react'
import PropTypes from 'prop-types'

const PageContent = () => {
  return <div>Page content</div>
}

PageContent.propTypes = {
  children: PropTypes.node
}

PageContent.displayName = 'PageContent'

export default PageContent