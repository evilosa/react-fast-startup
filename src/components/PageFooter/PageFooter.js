import React from 'react'
import PropTypes from 'prop-types'

const PageFooter = () => {
  return <div>Page footer</div>
}

PageFooter.propTypes = {
  children: PropTypes.node
}

PageFooter.displayName = 'PageFooter'

export default PageFooter