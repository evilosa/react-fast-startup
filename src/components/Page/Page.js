import * as React from 'react'
import style from './style';
import Radium from 'radium';

class Page extends React.Component {

  static defaultProps = {
    header: '',
  }

  _renderHeader = () => {
    const { header, onClose } = this.props;

    return (
      <div
        className='page-header'
        style={style.header}
      >
        {header && <div style={style.headerTitle}><h1>{header}</h1></div>}
        {onClose && <div style={style.headerButton}><button onClick={onClose}>X</button></div>}
      </div>
    )
  }

  _renderContent = content => (
    <div
      className='page-content'
      style={style.content}
    >
      {content}
    </div>
  )

  _renderFooter = footer => (
    <div
      className='page-footer'
      style={style.footer}
    >
      {footer}
    </div>
  )

  render() {
    const { children } = this.props
    let content = ''
    let footer = ''

    React.Children.forEach(children, (child) => {
      if (!child.type
        || (
          child.type.displayName !== 'PageContent'
          && child.type.displayName !== 'PageFooter'
      ))
        throw new Error('Child component should be instance of <PageContent/> or <PageFooter/>')

      const { children } = child.props

      switch (child.type.displayName) {
        case 'PageContent':
          content = children
          break
        case 'PageFooter':
          footer = children
          break
      }
    })

    return (
      <div
        className='page-root'
        style={style.root}
      >
        {this._renderHeader()}
        {this._renderContent(content)}
        {this._renderFooter(footer)}
      </div>
    )
  }
}

export default Radium(Page);