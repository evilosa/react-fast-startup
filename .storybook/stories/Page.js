import React from 'react';
import { withInfo } from '@storybook/addon-info';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Page from '../../src/components/Page';
import PageContent from '../../src/components/PageContent';
import PageFooter from '../../src/components/PageFooter';

storiesOf('Page', module)
  .add('default',
    withInfo({
      components: { Page },
      header: true,
      inline: true,
      text: 'Default page layout',
    })(() => (
      <Page
        header='Page header'
        onClose={action('clicked')}
      >
        <PageContent>
          <div>Content</div>
        </PageContent>
        <PageFooter>
          <div>Footer</div>
        </PageFooter>
      </Page>
  )));