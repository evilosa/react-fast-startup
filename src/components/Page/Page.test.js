import React from 'react'
import { shallow, mount } from 'enzyme'
import Page from './Page'
import PageContent from '../PageContent'
import PageFooter from '../PageFooter'

describe('Page', () => {
  it('renders properly', () => {
    const page = shallow(<Page/>)
    expect(page).toMatchSnapshot()
  })

  it('has correct default props', () => {
    const page = mount(<Page/>)
    const props = page.props()
    expect(props.header).toEqual('')
    expect(props.onClose).toBeUndefined()
  })

  describe('with header', () => {
    const page = shallow(
      <Page
        header='Test header'
      >
      </Page>
    )

    it('should render correct header', () => {
      expect(page.contains(<h1>Test header</h1>)).toBeTruthy()
    })
  })

  describe('onClose button', () => {
    it('should render and be able to click if provided', () => {
      const onClick = jest.fn();
      const page = shallow(
        <Page
          onClose={onClick}
        >
        </Page>
      )

      page.find('button').simulate('click')
      expect(onClick.mock.calls.length).toEqual(1)
    })

    it('should not be rendered if function is not provided', () => {
      const page = shallow(<Page/>)
      expect(page.find('button').length).toEqual(0)
    })
  })

  describe('with content', () => {
    const page = shallow(
      <Page>
        <PageContent>
          <div className='test-page-content'>Footer content</div>
        </PageContent>
      </Page>
    )

    it('should render empty footer', () => {
      expect(page.find('.page-footer').exists()).toBeTruthy()
    })

    it('should render content', () => {
      expect(page.find('.page-content').exists()).toBeTruthy()
    })

    it('should render content children', () => {
      expect(page.find('.test-page-content').exists()).toBeTruthy()
    })
  })

  describe('with footer', () => {
    const page = shallow(
      <Page>
        <PageFooter>
          <div className='test-footer-content'>Footer content</div>
        </PageFooter>
      </Page>
    )

    it('should render empty content', () => {
      expect(page.find('.page-content').exists()).toBeTruthy()
    })

    it('should render footer', () => {
      expect(page.find('.page-footer').exists()).toBeTruthy()
    })

    it('should render footer children', () => {
      expect(page.find('.test-footer-content').exists()).toBeTruthy()
    })
  })

  describe('with wrong children', () => {
    const component =
      <Page>
        <div>Wrong children</div>
      </Page>

    it('should throw error', () => {
      expect(() => shallow(component)).toThrow();
    })
  })
})