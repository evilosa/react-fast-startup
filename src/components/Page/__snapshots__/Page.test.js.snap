// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders properly 1`] = `
<div
  className="page-root"
>
  <div
    className="page-header"
  />
  <div
    className="page-content"
  />
  <div
    className="page-footer"
  />
</div>
`;
